pkg/distributor/query.go

```diff
@@ -2,6 +2,7 @@
 // Provenance-includes-location: https://github.com/cortexproject/cortex/blob/master/pkg/distributor/query.go
 // Provenance-includes-license: Apache-2.0
 // Provenance-includes-copyright: The Cortex Authors.
+// Provenance-includes-location: https://github.com/grafana/loki/blob/main/pkg/util/loser/tree.go
```

Comment:
maybe this tree can go in dskit

Reply:
This implementation of the loser tree is specialised compared to the one in dskit, and I don't think it's generally useful - perhaps we can extract it when we have a second place that needs it?

Reply:
I haven't compared the two myself. What's the specialization here? Can we make the dskit one customizable so that we can reuse it here as well?

Reply:
There are two main differences compared to the implementation already in dskit (which is based on what Loki has):

* the implementation here returns the index of the series within the ingester, whereas dskit's implementation does not
* the implementation in dskit assumes the elements of the slices being merged satisfy [`constraints.Ordered`](https://pkg.go.dev/golang.org/x/exp/constraints#Ordered), but that is only satisfied by primitive types

Reply:
Low priority, but could be interesting to see if we can make the loser tree in dskit more extensible so that we can use it for our purposes here too.